### PR TITLE
feat(artifacts): Support embedded/base64 artifact type

### PR DESCRIPTION
**Pull request by sbwsg**
fixes https://github.com/spinnaker/spinnaker/issues/2526

default_artifact_base64

With invalid input:

base64_latin1_error

### CODE REVIEW

File: app/scripts/modules/core/src/pipeline/config/triggers/artifacts/artifact.component.ts [hunk 1]
```diff
@@ -52,7 +52,7 @@ class ArtifactCtrl implements IController {
 
     if (artifactKindConfig.length) {
       const config = artifactKindConfig[0];
-      const template: Element = config.template as any;
+      const template: string = config.template as any;
```
Comment (danielpeach) [inline app/scripts/modules/core/src/pipeline/config/triggers/artifacts/artifact.component.ts]: Isn't the type of `template` on `IArtifactKindConfig` a `string`? Should be able to drop `as any`.

File: app/scripts/modules/core/src/pipeline/config/triggers/artifacts/base64/base64.artifact.ts [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+      Name
+    </label>
+    <div class="col-md-8">
+      <input type="text"
+             placeholder="base64-artifact"
+             class="form-control input-sm"
+             ng-model="ctrl.artifact.name" />
+    </div>
+  </div>
+  <div ng-if="ctrl.encodeDecodeError != null" class="form-group row">
```
Comment (danielpeach) [inline app/scripts/modules/core/src/pipeline/config/triggers/artifacts/base64/base64.artifact.ts]: In most of the spots where we surface this kind of error (see the Pipeline JSON edit modal), we put the error below the text box.

**Comment by danielpeach** [comment]
Nice. 

- Is there any reasonable way to drop the "Failed to execute 'btoa' on 'Window'" part of the error message without parsing the error message? Users probably don't care how we converted their string to base64.
- Someone can disagree with me, but I wonder how useful the Base64 text box is - it's probably just noise in most cases, and users might be better served by a tooltip and/or a copy-to-clipboard icon.

File: app/scripts/modules/core/src/pipeline/config/triggers/artifacts/base64/base64.artifact.ts [hunk 3]
```diff
@@ -0,0 +1,87 @@
+import { module } from 'angular';
+
+import { PIPELINE_CONFIG_PROVIDER } from 'core/pipeline/config/pipelineConfigProvider';
+import { IArtifact } from 'core/domain/IArtifact';
+import { PipelineConfigProvider } from 'core/pipeline';
+
+import './base64.artifact.less';
+
+export const BASE64_ARTIFACT = 'spinnaker.core.pipeline.trigger.github.artifact';
```
[the remark this answers is not in the document]
Comment (sbwsg) [reply app/scripts/modules/core/src/pipeline/config/triggers/artifacts/base64/base64.artifact.ts]: whoops, thankyou!

**Comment by sbwsg** [comment]
@ezimanyi yes indeed, that work was already done here: https://github.com/spinnaker/clouddriver/pull/2416/files

@dpeach for 1) I've added an explicit error message for when a char is outside the latin1 range so that it wont have the "Failed to executed 'btoa' on 'window'" part.  I expect that will be the most common error when converting between base64 / decoded content.  All other errors will be printed to screen with the entire error message.  Short of parsing it I can't think of a way to remove that string prefix.  And for 2) I agree about the lack of utility of the base64 text content being in the UI.  I've replaced the `<pre>` with a copy-to-clipboard icon.

**Comment by sbwsg** [comment]
@lwander I've enabled the artifact as a matchArtifact.

File: app/scripts/modules/core/src/pipeline/config/triggers/artifacts/base64/base64.artifact.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    <div class="col-md-8">
+      <input type="text"
+             placeholder="base64-artifact"
+             class="form-control input-sm"
+             ng-model="ctrl.artifact.name" />
+    </div>
+  </div>
+  <div ng-if="ctrl.encodeDecodeError" class="form-group row">
+    <div class="col-md-12 error-message">
+      Error encoding/decoding artifact content: {{ ctrl.encodeDecodeError }}
```
Comment (lwander) [inline app/scripts/modules/core/src/pipeline/config/triggers/artifacts/base64/base64.artifact.ts]: Why does the match artifact have encoding errors?

**Pull request merged**